src/extensions/extensionManager.ts

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2025, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import * as vscode from 'vscode';
+import {
+  ExtensionState,
+  normalizeWaitOptions,
+  ServiceProviders,
+  serviceTypeToProvider,
+  ServiceWaitResult,
+  WaitOptions
+} from '../types';
+// Below import has to be required for bundling
+// eslint-disable-next-line @typescript-eslint/no-var-requires, @typescript-eslint/no-unsafe-assignment
+const AsyncLock = require('async-lock');
+
+export default class ExtensionManager {
+  private static instance: ExtensionManager;
+  private readonly extensionStates: Map<
+    ServiceProviders,
+    vscode.Extension<unknown> | undefined
+  >;
+  private lock = new AsyncLock();
+
+  private constructor() {
+    this.extensionStates = new Map<
+      ServiceProviders,
+      vscode.Extension<unknown> | undefined
+    >();
+    this.setupExtensionChangeHandler();
+  }
+
+  public static async getInstance(): Promise<ExtensionManager> {
+    if (!ExtensionManager.instance) {
+      ExtensionManager.instance = new ExtensionManager();
+      await ExtensionManager.instance.initializeExtensionStates();
+    }
+    return ExtensionManager.instance;
+  }
+
+  protected async initializeExtensionStates(): Promise<void> {
+    await this.lock.acquire(this.extensionStates, () => {
+      this.extensionStates.clear();
+      const processedProviders = new Set(Object.values(serviceTypeToProvider));
+
+      processedProviders.forEach((provider) => {
+        const extension = vscode.extensions.getExtension(provider);
+        this.extensionStates.set(provider, extension);
+        processedProviders.add(provider);
+      });
+    });
+  }
+
+  private handleExtensionChange() {
+    Object.values(serviceTypeToProvider).forEach((provider) => {
+      const extension = vscode.extensions.getExtension(provider);
+      this.extensionStates.set(provider, extension);
+    });
+  }
+
+  public isExtensionActive(provider: ServiceProviders): boolean {
+    const extension = this.extensionStates.get(provider);
+    return extension?.isActive || false;
+  }
+
+  public getExtensionStates(): Map<
+    ServiceProviders,
+    vscode.Extension<unknown> | undefined
+  > {
+    return this.extensionStates;
+  }
+
+  public getExtensionState(
+    provider: ServiceProviders
+  ): vscode.Extension<unknown> {
+    return this.extensionStates.get(provider);
+  }
+
+  public static async waitForExtensionToBecomeActive(
+    provider: ServiceProviders,
+    options: WaitOptions = {
+      timeout: 60_000,
```

Comment:
I've never seen an underscore in an integer value like that. It's just to increase readability, right?

Reply:
Yes, improves readability.